nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -365,6 +365,10 @@ in
           assertion = config.networking.hostId != null;
           message = "ZFS requires networking.hostId to be set";
         }
+        {
+          assertion = !config.boot.loader.grub.enable || config.boot.loader.grub.copyKernels;
+          message = "ZFS requires boot.loader.grub.copyKernels to be enabled";
+        }
```

Comment:
I don't think this needs to be an assertion. Just setting `copyKernels` to `true` if grub is enabled shouldn't be a problem.

Reply:
It could be potentially a breaking change for existing users with small boot partitions - do you think it's okay?

As a kinda data point, I have only 110M used on /boot* with `boot.loader.grub.configurationLimit = 50;` (while default is 100). So, it's a very small amount of data.

Reply:
Ah yeah an assertion sounds good then. Though I'd change the condition to
```nix
assertion = config.boot.loader.grub.enable -> config.boot.loader.grub.copyKernels
```

Reply:
@Infinisil @Ma27 
It seems that due to the zfs enabled by default, there is no way to do a safe assertion (or I understand it in a wrong way). Right now, this assertion will basically mean that anyone who's using grub will need copyKernels to be enabled as well.

On the other hand, I searched a bit dipper and found that copyKernels will be effectively forced if  /nix/store and boot path are on different filesystems [0]. Seems like it should be enough, but it makes copyKernels kind of impure.

[0] https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/loader/grub/install-grub.pl#L83-L88

Reply:
Would it work to check the actual `fsType` of `/boot` (or `/` if there is no separate `/boot`)? 
```
(config.fileSystems."/boot" or config.fileSystems."/").fsType == "zfs"
```

Reply:
@chkno that's a good idea. Probably checking the root fs is enough as if it's zfs we should be already copying kernels as well as /boot doesn't always exist (for example, I have /boot[1-6]). I'll try a few different options and see how it'll work.